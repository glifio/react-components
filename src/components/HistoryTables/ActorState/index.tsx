import PropTypes from 'prop-types'
import { useEffect, useMemo } from 'react'
import { FilecoinNumber } from '@glif/filecoin-number'
import {
  DataTypeMap,
  getActorName,
  describeActorState
} from '@glif/filecoin-actor-utils'

import { useAddressQuery } from '../../../generated/graphql'
import { makeFriendlyBalance } from '../../../utils/makeFriendlyBalance'
import { useStateReadState } from '../../../utils/useStateReadState'
import { useMsigGetAvailableBalance } from '../../../utils/useMsigGetAvailableBalance'
import convertAddrToPrefix from '../../../utils/convertAddrToPrefix'
import {
  Lines,
  Line,
  NullishLine,
  PageTitle,
  CollapsableLines
} from '../../Layout'
import { DetailCaption } from '../detail'
import {
  useEnvironment,
  useLogger
} from '../../../services/EnvironmentProvider'
import { BaseTypeObjLines, DataTypeMapLines } from '../../Layout/DataTypes'
import { isEthAddress, isFilAddress } from '../../../utils/isAddress'
import {
  decode,
  ethAddressFromDelegated,
  newDelegatedEthAddress,
  Protocol
} from '@glif/filecoin-address'

export const ActorState = ({ address: addressProp }: ActorStateProps) => {
  const logger = useLogger()
  const { coinType, networkName } = useEnvironment()

  // Ensure network cointype and address type for address
  const address = useMemo<string>(() => {
    if (isFilAddress(addressProp)) {
      return convertAddrToPrefix(addressProp, coinType)
    } else if (isEthAddress(addressProp)) {
      return newDelegatedEthAddress(addressProp, coinType).toString()
    }
    return ''
  }, [addressProp, coinType])

  const subAddr = useMemo<string | null>(() => {
    if (address && decode(address).protocol() === Protocol.DELEGATED) {
      return ethAddressFromDelegated(address)
    }

    return null
  }, [address])

  // Load the actor state
  const {
    data: actorData,
    loading: actorLoading,
    notFound: actorNotFound,
    error: actorError
  } = useStateReadState(address)

  // Load the address
  const {
    data: addressData,
    loading: addressLoading,
    error: addressError
  } = useAddressQuery({
    variables: { address }
  })

  // Get the actor name from the actor code
  const actorName = useMemo<string | null>(
    () => (actorData ? getActorName(actorData.Code, networkName) : null),
    [actorData, networkName]
  )

  // Capitalize the actor name for display
  const actorNameCapitalized = useMemo<string | null>(
    () =>
      actorName ? actorName.charAt(0).toUpperCase() + actorName.slice(1) : null,
    [actorName]
  )

  // Get actor state with descriptors
  const describedState = useMemo<DataTypeMap | null>(() => {
    try {
      return actorName && actorData?.State
        ? describeActorState(actorName, actorData.State)
        : null
    } catch (e) {
      logger.error(
        `Failed to describe actor state for network: ${networkName}, address: ${address}, with message: ${e.message}`
      )
      return null
    }
  }, [actorName, actorData, address, networkName, logger])

  // Load the available balance for multisig actors
  const hasAvailableBalance = actorName && actorName.includes('multisig')
  const {
    availableBalance,
    loading: availableBalanceLoading,
    error: availableBalanceError
  } = useMsigGetAvailableBalance(hasAvailableBalance ? address : '')

  // Cache friendly balance
  const friendlyBalance = useMemo<string>(() => {
    if (!actorData) return ''
    const balance = new FilecoinNumber(actorData.Balance, 'attofil')
    return makeFriendlyBalance(balance, 6)
  }, [actorData])

  // Log actor state errors
  useEffect(() => actorError && logger.error(actorError), [actorError, logger])

  // Log address errors
  useEffect(
    () => addressError && logger.error(addressError),
    [addressError, logger]
  )

  // Log available balance errors
  useEffect(
    () => availableBalanceError && logger.error(availableBalanceError),
    [availableBalanceError, logger]
  )

  // Actor state or address loading
  const loading = useMemo<boolean>(() => {
    return actorLoading || addressLoading
  }, [actorLoading, addressLoading])

  // Actor state or address error
  const error = useMemo<Error | null>(() => {
    return actorError || addressError || null
  }, [actorError, addressError])

  return (
    <>
      <PageTitle>Actor Overview</PageTitle>
      <hr />
      <DetailCaption
        name='Actor Overview'
        infoMsg={actorNotFound ? `Actor not found: ${address}` : ''}
        loadingMsg='Locating this actor on the blockchain...'
        loading={loading}
        error={error}
      />
      {!loading && !error && !actorNotFound && (
        <Lines>
          {addressData?.address.robust && (
            <Line label='Robust address'>{addressData?.address.robust}</Line>
          )}
          {!!subAddr && <Line label='ETH address'>{subAddr}</Line>}
          {addressData?.address.id && (
            <Line label='ID address'>{addressData?.address.id}</Line>
          )}
          <Line label='Actor name'>{actorNameCapitalized || 'Unknown'}</Line>
          <Line label='Actor code'>{actorData.Code['/']}</Line>
          <Line label='Balance'>{friendlyBalance} FIL</Line>
          {hasAvailableBalance && (
            <Line label='Available Balance'>
              {availableBalanceError ? (
                <>Failed to load</>
              ) : availableBalanceLoading ? (
                <>Loading...</>
              ) : availableBalance ? (
                <>{availableBalance.toFil()} FIL</>
              ) : (
                <></>
              )}
            </Line>
          )}
          {actorData.State ? (
            <CollapsableLines label='State' toggleName='actor state'>
              {describedState ? (
                <DataTypeMapLines depth={1} dataTypeMap={describedState} />
              ) : (
                <BaseTypeObjLines depth={1} data={actorData.State} />
              )}
            </CollapsableLines>
          ) : (
            <NullishLine label='State' />
          )}
        </Lines>
      )}
    </>
  )
}

export interface ActorStateProps {
  address: string
}

ActorState.propTypes = {
  address: PropTypes.string
}
