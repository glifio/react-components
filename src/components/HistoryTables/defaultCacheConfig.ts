import { InMemoryCacheConfig } from '@apollo/client'
import Logger from '../../utils/logger'
import { removeMessageDups } from './utils'

// The params field is expected to be a JSON string
// or null. Both are safe to pass to JSON parse. The
// result should be a valid javascript object or null.
const parseParams = (_: any, incoming: any) => {
  try {
    if (incoming) {
      let params = JSON.parse(incoming)
      if (typeof params.Value === 'object' && !!params.Value.Int) {
        params.Value = params.Value.Int
      }
      return params
    }
    return null
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    // Logger.error(`Failed to parse MessageConfirmed.params: ${msg}`)
    return null
  }
}

export const defaultMessageHistoryClientCacheConfig: InMemoryCacheConfig = {
  typePolicies: {
    Query: {
      fields: {
        messages: {
          // Cache separate results based on address passed as arg
          keyArgs: ['address'],
          // Concatenate the incoming list items with
          // the existing list items.
          merge(existing = [], incoming = [], { args }) {
            return removeMessageDups(
              existing ? [...existing] : [],
              incoming ? [...incoming] : [],
              args as unknown as { limit: number; offset: number }
            )
          }
        },
        message: {
          read(_, { args, toReference }) {
            return toReference({
              __typename: 'Message',
              cid: args.cid
            })
          }
        }
      }
    },
    Block: {
      keyFields: ['Cid']
    },
    Message: {
      keyFields: ['cid'],
      fields: {
        params: {
          merge: parseParams
        }
      }
    },
    Address: {
      keyFields: ['robust', 'id']
    },
    MessagePending: {
      keyFields: ['cid'],
      fields: {
        params: {
          merge: parseParams
        }
      }
    },
    MsigTransaction: {
      keyFields: ['id'],
      fields: {
        params: {
          merge: parseParams
        }
      }
    }
  }
}
