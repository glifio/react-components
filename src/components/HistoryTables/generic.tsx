import React from 'react'
import PropTypes from 'prop-types'
import Box from '../Box'

export const Badge = ({ color, children }: BadgeProps) => (
  <Box
    display='flex'
    alignItems='center'
    justifyContent='center'
    gridGap='0.75em'
    height='2em'
    lineHeight='2em'
    borderRadius='1em'
    px='1em'
    bg={`${color}.light`}
    color={`${color}.medium`}
    style={{ whiteSpace: 'nowrap' }}
    textAlign='center'
  >
    {children}
  </Box>
)

type BadgeProps = {
  color: 'purple' | 'green' | 'yellow' | 'red'
  children: React.ReactNode
}

Badge.propTypes = {
  color: PropTypes.oneOf(['purple', 'green', 'yellow', 'red']).isRequired,
  children: PropTypes.node.isRequired
}
