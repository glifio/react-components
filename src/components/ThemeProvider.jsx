import PropTypes from 'prop-types'
import { ThemeProvider as StyledThemeProvider } from 'styled-components'
import { theme as defaultTheme } from './theme'

const ThemeProvider = ({ children, theme, ...rest }) => {
  return (
    <StyledThemeProvider theme={theme} {...rest}>
      {children}
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
}

ThemeProvider.defaultProps = {
  theme: defaultTheme,
  children: <></>
}

ThemeProvider.displayName = 'ThemeProvider'

export default ThemeProvider
