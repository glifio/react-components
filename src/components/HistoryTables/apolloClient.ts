import { ApolloClient, InMemoryCache, HttpLink, split } from '@apollo/client'
import { getMainDefinition } from '@apollo/client/utilities'
import { WebSocketLink } from '@apollo/client/link/ws'
import { defaultMessageHistoryClientCacheConfig } from './defaultCacheConfig'
import { networks } from '../../services/EnvironmentProvider'

const BASE_URL = networks.calibration.graphUrl

const httpLink = new HttpLink({
  uri: `https://${BASE_URL}`
})

const wsLink = process.browser
  ? new WebSocketLink({
      uri: `wss://${BASE_URL}`,
      options: {
        reconnect: true,
        lazy: true
      }
    })
  : null

// The split function takes three parameters:
//
// * A function that's called for each operation to execute
// * The Link to use for an operation if the function returns a "truthy" value
// * The Link to use for an operation if the function returns a "falsy" value
const link = process.browser //only create the split in the browser
  ? split(
      ({ query }) => {
        const definition = getMainDefinition(query)
        return (
          definition.kind === 'OperationDefinition' &&
          definition.operation === 'subscription'
        )
      },
      wsLink,
      httpLink
    )
  : httpLink

export const client = new ApolloClient({
  link,
  cache: new InMemoryCache({ ...defaultMessageHistoryClientCacheConfig })
})
