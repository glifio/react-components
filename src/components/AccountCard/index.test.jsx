/* eslint-disable import/first */
/* eslint-disable import/newline-after-import */
import { render, screen, act, fireEvent } from '@testing-library/react'
import { ApolloProvider } from '@apollo/client'
import theme from '../theme'
import ThemeProvider from '../ThemeProvider'
import { AccountCard } from '.'
jest.mock('../../utils/copyToClipboard')
import copyToClipboard from '../../utils/copyToClipboard'
import { Base } from './index.stories'
import { client } from '../HistoryTables/apolloClient'
import { TestEnvironment } from '../../test-utils/TestEnvironment'

describe('AccountCard', () => {
  test('renders the story', () => {
    const { container } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <Base {...Base.args} />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders the card', () => {
    const { container } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={() => {}}
              color='purple'
              address='t0123456789'
              walletType='LEDGER'
              onShowOnLedger={() => {}}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders the card CREATE_MNEMONIC', () => {
    const { container } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={() => {}}
              color='purple'
              address='t0123456789'
              walletType='CREATE_MNEMONIC'
              onShowOnLedger={() => {}}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders the address', () => {
    render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={() => {}}
              color='purple'
              address='t0123'
              walletType='LEDGER'
              onShowOnLedger={() => {}}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    expect(screen.getByText('t0123', { exact: false })).toBeInTheDocument()
    expect(screen.getByText('Your Address')).toBeInTheDocument()
  })

  test('clicking "Switch" calls onAccountSwitch', () => {
    const mockOnAccountSwitch = jest.fn()
    const { getByText } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={mockOnAccountSwitch}
              color='purple'
              address='t0123'
              walletType='LEDGER'
              onShowOnLedger={() => {}}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    act(() => {
      fireEvent.click(getByText('Switch'))
    })

    expect(mockOnAccountSwitch).toHaveBeenCalled()
  })

  test('clicking "Show on Device" calls onShowOnLedger', () => {
    const mockOnShowOnLedger = jest.fn()
    const { getByText } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={() => {}}
              color='purple'
              address='t0123'
              walletType='LEDGER'
              onShowOnLedger={mockOnShowOnLedger}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    act(() => {
      fireEvent.click(getByText('Show on Device'))
    })

    expect(mockOnShowOnLedger).toHaveBeenCalled()
  })

  test('clicking "Copy" calls copy', async () => {
    const mockCopyToClipboard = jest.fn(() => Promise.resolve('yo'))
    copyToClipboard.mockImplementationOnce(mockCopyToClipboard)

    const { getByText } = render(
      <TestEnvironment>
        <ThemeProvider theme={theme}>
          <ApolloProvider client={client}>
            <AccountCard
              onAccountSwitch={() => {}}
              color='purple'
              address='t0123'
              walletType='LEDGER'
              onShowOnLedger={() => {}}
              ledgerBusy={false}
              mb={2}
            />
          </ApolloProvider>
        </ThemeProvider>
      </TestEnvironment>
    )

    act(() => {
      fireEvent.click(getByText('Copy'))
    })

    expect(mockCopyToClipboard).toHaveBeenCalled()
    expect(screen.getByText('Copied')).toBeInTheDocument()
  })
})
