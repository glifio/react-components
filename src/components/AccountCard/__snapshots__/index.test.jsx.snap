// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCard renders the card 1`] = `
.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  color: var(--white);
  border-color: var(--white);
}

.c8 svg path {
  stroke: var(--black);
}

.c8:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c8:hover svg path {
  stroke: var(--purple-medium);
}

.c8:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c8:active svg path {
  stroke: var(--white) !important;
}

.c8 svg path {
  stroke: var(--white);
}

.c6 {
  width: 24;
  height: 24;
}

.c0 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c0 > *:first-child {
  margin-top: 0;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

.c0 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c0 > h2:first-child {
  margin: 0;
}

.c0 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 {
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 hr {
  border-color: var(--white);
}

.c3 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c3 p {
  margin: 0;
  line-height: 1.5;
}

.c5 {
  position: relative;
  cursor: pointer;
}

.c5 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c5 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100% + 0.25em);
  opacity: 0;
  color: var(--white);
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
}

.c5:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c5:hover .copy-text {
  opacity: 1;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25em;
  line-height: 1.5;
}

.c4 > a {
  color: var(--white);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 > a:hover {
  color: var(--white);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c4 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c4 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c7 > p,
.c7 > hr {
  margin: 0.5em 0;
}

.c7 > p[role='button'] {
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 18em;
  height: 18em;
}

.c2 h3 {
  margin-top: 0;
  text-align: left;
}

.c2 h4 {
  color: var(--purple-light);
  text-align: left;
}

.c2 .buttons button:hover {
  color: var(--purple-medium);
  border-color: var(--white);
  background-color: var(--white);
}

.c2 .buttons button:active {
  color: var(--purple-medium) !important;
  border-color: var(--purple-light) !important;
  background-color: var(--purple-light) !important;
}

<div
  class="c0 c1 c2"
>
  <div>
    <h3>
      Account
    </h3>
    <hr />
    <div
      class="c3"
    >
      <h4>
        Your Address
      </h4>
      <div
        class="c4"
        color="var(--white)"
      >
        <a
          href="https://explorer-calibration.glif.link/actor/?address=t0123456789"
          rel="noreferrer noopener"
          target="_blank"
        >
          t0123456789
        </a>
        <span
          class="c5"
          color="var(--white)"
          role="button"
        >
          <svg
            class="c6"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#1A0066"
              height="12"
              width="12"
              x="8"
              y="4"
            />
            <rect
              fill="#E0D7FE"
              height="12"
              width="12"
              x="4"
              y="8"
            />
          </svg>
          <span
            class="copy-text"
          >
            Copy
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="buttons"
  >
    <div
      class="c7"
    >
      <button
        class="c8"
      >
        Show on Device
      </button>
      <button
        class="c8"
      >
        Switch
      </button>
    </div>
  </div>
</div>
`;

exports[`AccountCard renders the card CREATE_MNEMONIC 1`] = `
.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  color: var(--white);
  border-color: var(--white);
}

.c8 svg path {
  stroke: var(--black);
}

.c8:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c8:hover svg path {
  stroke: var(--purple-medium);
}

.c8:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c8:active svg path {
  stroke: var(--white) !important;
}

.c8 svg path {
  stroke: var(--white);
}

.c6 {
  width: 24;
  height: 24;
}

.c0 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c0 > *:first-child {
  margin-top: 0;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

.c0 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c0 > h2:first-child {
  margin: 0;
}

.c0 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 {
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 hr {
  border-color: var(--white);
}

.c3 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c3 p {
  margin: 0;
  line-height: 1.5;
}

.c5 {
  position: relative;
  cursor: pointer;
}

.c5 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c5 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100% + 0.25em);
  opacity: 0;
  color: var(--white);
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
}

.c5:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c5:hover .copy-text {
  opacity: 1;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25em;
  line-height: 1.5;
}

.c4 > a {
  color: var(--white);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 > a:hover {
  color: var(--white);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c4 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c4 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c7 > p,
.c7 > hr {
  margin: 0.5em 0;
}

.c7 > p[role='button'] {
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 18em;
  height: 18em;
}

.c2 h3 {
  margin-top: 0;
  text-align: left;
}

.c2 h4 {
  color: var(--purple-light);
  text-align: left;
}

.c2 .buttons button:hover {
  color: var(--purple-medium);
  border-color: var(--white);
  background-color: var(--white);
}

.c2 .buttons button:active {
  color: var(--purple-medium) !important;
  border-color: var(--purple-light) !important;
  background-color: var(--purple-light) !important;
}

<div
  class="c0 c1 c2"
>
  <div>
    <h3>
      Account
    </h3>
    <hr />
    <div
      class="c3"
    >
      <h4>
        Your Address
      </h4>
      <div
        class="c4"
        color="var(--white)"
      >
        <a
          href="https://explorer-calibration.glif.link/actor/?address=t0123456789"
          rel="noreferrer noopener"
          target="_blank"
        >
          t0123456789
        </a>
        <span
          class="c5"
          color="var(--white)"
          role="button"
        >
          <svg
            class="c6"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#1A0066"
              height="12"
              width="12"
              x="8"
              y="4"
            />
            <rect
              fill="#E0D7FE"
              height="12"
              width="12"
              x="4"
              y="8"
            />
          </svg>
          <span
            class="copy-text"
          >
            Copy
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="buttons"
  >
    <div
      class="c7"
    >
      <button
        class="c8"
      >
        Switch
      </button>
    </div>
  </div>
</div>
`;

exports[`AccountCard renders the story 1`] = `
.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  color: var(--white);
  border-color: var(--white);
}

.c8 svg path {
  stroke: var(--black);
}

.c8:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c8:hover svg path {
  stroke: var(--purple-medium);
}

.c8:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c8:active svg path {
  stroke: var(--white) !important;
}

.c8 svg path {
  stroke: var(--white);
}

.c6 {
  width: 24;
  height: 24;
}

.c0 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c0 > *:first-child {
  margin-top: 0;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

.c0 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c0 > h2:first-child {
  margin: 0;
}

.c0 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 {
  background-color: var(--purple-medium);
  color: var(--white);
}

.c1 hr {
  border-color: var(--white);
}

.c3 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c3 p {
  margin: 0;
  line-height: 1.5;
}

.c5 {
  position: relative;
  cursor: pointer;
}

.c5 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c5 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100% + 0.25em);
  opacity: 0;
  color: var(--white);
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
}

.c5:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c5:hover .copy-text {
  opacity: 1;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25em;
  line-height: 1.5;
}

.c4 > a {
  color: var(--white);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 > a:hover {
  color: var(--white);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c4 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c4 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c7 > p,
.c7 > hr {
  margin: 0.5em 0;
}

.c7 > p[role='button'] {
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 18em;
  height: 18em;
}

.c2 h3 {
  margin-top: 0;
  text-align: left;
}

.c2 h4 {
  color: var(--purple-light);
  text-align: left;
}

.c2 .buttons button:hover {
  color: var(--purple-medium);
  border-color: var(--white);
  background-color: var(--white);
}

.c2 .buttons button:active {
  color: var(--purple-medium) !important;
  border-color: var(--purple-light) !important;
  background-color: var(--purple-light) !important;
}

<div
  class="c0 c1 c2"
>
  <div>
    <h3>
      Account
    </h3>
    <hr />
    <div
      class="c3"
    >
      <h4>
        Your Address
      </h4>
      <div
        class="c4"
        color="var(--white)"
      >
        <a
          href="https://explorer-calibration.glif.link/actor/?address=t0123456789"
          rel="noreferrer noopener"
          target="_blank"
        >
          t0123456789
        </a>
        <span
          class="c5"
          color="var(--white)"
          role="button"
        >
          <svg
            class="c6"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#1A0066"
              height="12"
              width="12"
              x="8"
              y="4"
            />
            <rect
              fill="#E0D7FE"
              height="12"
              width="12"
              x="4"
              y="8"
            />
          </svg>
          <span
            class="copy-text"
          >
            Copy
          </span>
        </span>
      </div>
    </div>
  </div>
  <div
    class="buttons"
  >
    <div
      class="c7"
    >
      <button
        class="c8"
      >
        Show on Device
      </button>
      <button
        class="c8"
      >
        Switch
      </button>
    </div>
  </div>
</div>
`;
